src/rgw/rgw_dedup.cc

```diff
@@ -0,0 +1,2363 @@
+      num_work_shards = work_distrib(gen);
+    }
+    while (num_md5_shards == 0) {
+      num_md5_shards = md5_distrib(gen);
+    }
+    ldpp_dout(dpp, 10) << __func__ << "::RAND num_work_shards=" << num_work_shards
+                       << "::num_md5_shards=" << num_md5_shards << dendl;
+    int ret = d_cluster.reset(store, d_dedup_cluster_ioctx, p_epoch,
+                              num_work_shards, num_md5_shards);
+#else
+    int ret = d_cluster.reset(store, d_dedup_cluster_ioctx, p_epoch,
+                              MAXX_WORK_SHARD, MAXX_MD5_SHARD);
+#endif
+    if (ret != 0) {
+      derr << __func__ << "::ERR: failed cluster.init()" << dendl;
+      return -1;
+    }
+
+    ldpp_dout(dpp, 10) <<__func__ << "::" << *p_epoch << dendl;
+    d_ctl.dedup_type = p_epoch->dedup_type;
+#ifdef FULL_DEDUP_SUPPORT
+    ceph_assert(d_ctl.dedup_type == dedup_req_type_t::DEDUP_TYPE_FULL ||
+                d_ctl.dedup_type == dedup_req_type_t::DEDUP_TYPE_DRY_RUN);
+#else
+    ceph_assert(d_ctl.dedup_type == dedup_req_type_t::DEDUP_TYPE_DRY_RUN);
+#endif
+    ldpp_dout(dpp, 10) << __func__ << "::" << d_ctl.dedup_type << dendl;
+
+    return 0;
+  }
+#endif
+  //---------------------------------------------------------------------------
+  int Background::watch_reload(const DoutPrefixProvider* dpp)
+  {
+    if (!d_dedup_cluster_ioctx.is_valid()) {
+      ldpp_dout(dpp, 1) << __func__
+                        << "::ERR: invalid pool handler (missing pool)" << dendl;
+      return -ENOENT;
+    }
+    const std::string & oid = DEDUP_WATCH_OBJ;
+    // create the object to watch (object may already exist)
+    bool exclusive = true;
+    int ret = d_dedup_cluster_ioctx.create(oid, exclusive);
+    if (ret >= 0) {
+      ldpp_dout(dpp, 5) << __func__ << "::" << oid << " was created!" << dendl;
+    }
+    else if (ret == -EEXIST) {
+      ldpp_dout(dpp, 5) << __func__ << "::"<< oid << " exists" << dendl;
+    }
+    else {
+      ldpp_dout(dpp, 1) << __func__ << "::ERR: failed ioctx.create(" << oid
+                        << ") with: " << ret << "::" << cpp_strerror(ret) << dendl;
+      return ret;
+    }
+
+    ret = d_dedup_cluster_ioctx.watch2(oid, &d_watch_handle, &d_watcher_ctx);
+    if (ret < 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::ERR: failed watch2() " << oid
+                        << ". error: " << cpp_strerror(ret) << dendl;
+      return ret;
+    }
+    ldpp_dout(dpp, 10) << __func__ << "::Started watching for reloads of  "
+                       << oid << ", handle=" << d_watch_handle << dendl;
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  int Background::unwatch_reload(const DoutPrefixProvider* dpp)
+  {
+    if (d_watch_handle == 0) {
+      // nothing to unwatch
+      return 0;
+    }
+
+    if (!d_dedup_cluster_ioctx.is_valid()) {
+      ldpp_dout(dpp, 1) << __func__
+                        << "::ERR: invalid pool handler (missing pool)" << dendl;
```

Comment:
Here and elsewhere, isn't ERROR best handled by the logging subsystem?  We're passing it a severity, no?  It will add DBG/WRN/ERR to the messages implicitly?

Reply:
I don't understand this comment

Reply:
The logging subsystem has no concept of INFO/WARN/ERROR and instead is using numeric values from 0-255.
I need a way to quickly search the logs for warning/errors reported by my code